File: pkg/security/secl/compiler/eval/pattern.go
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2016-present Datadog, Inc.

// Package eval holds eval related files
package eval

import (
	"strings"
)

func nextSegment(str string) (bool, string, int) {
	var inSegment bool
	var start, end int

	var star bool
	if str[0] == '*' {
		star = true
	}

	for i, c := range []byte(str) {
		if c != '*' {
			if !inSegment {
				start = i
				inSegment = true
			}
			end = i
		} else if inSegment {
			break
		}
	}

	if star && start == 0 {
		return star, "", 1
	}

	end++

	return star, str[start:end], end
}

func index(s, subtr string, caseInsensitive bool) int {
	if caseInsensitive {
		s = strings.ToLower(s)
		subtr = strings.ToLower(subtr)
	}
	return strings.Index(s, subtr)
}

func hasPrefix(s, prefix string, caseInsensitive bool) bool {
	if caseInsensitive {
		s = strings.ToLower(s)
		prefix = strings.ToLower(prefix)
	}
	return strings.HasPrefix(s, prefix)
}

// PatternMatches matches a pattern against a string
func PatternMatches(pattern string, str string, caseInsensitive bool) bool {
	patternElem := newPatternElement(pattern)
	return PatternMatchesWithSegments(patternElem, str, caseInsensitive)
}

// PatternMatchesWithSegments matches a pattern against a string
func PatternMatchesWithSegments(patternElem patternElement, str string, caseInsensitive bool) bool {
	if patternElem.pattern == "*" {
		return true
	}

	if len(patternElem.pattern) == 0 {
		return len(str) == 0
	}

	for _, seg := range patternElem.segments {
		if seg.star {
			// there is no pattern to match after the last star
			if len(seg.segment) == 0 {
				return true
			}

			index := index(str, seg.segment, caseInsensitive)
			if index == -1 {
				return false
			}
			str = str[index+len(seg.segment):]
		} else {
			if !hasPrefix(str, seg.segment, caseInsensitive) {
				return false
			}
			str = str[len(seg.segment):]
		}
	}

	// return false if there is still some str to match
	return len(str) == 0
}

type patternElement struct {
	pattern  string
	segments []patternSegment
}

type patternSegment struct {
	star    bool
	segment string
}

func newPatternElement(element string) patternElement {
	el := patternElement{
		pattern: element,
	}

	for len(element) > 0 {
		star, segment, nextIndex := nextSegment(element)
		element = element[nextIndex:]
		el.segments = append(el.segments, patternSegment{star, segment})
	}

	return el
}

File: pkg/discovery/module/rust/src/service_name/spring/pattern.rs
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2025-present Datadog, Inc.

/// Extract the longest path prefix without wildcards
pub fn longest_path_prefix(pattern: &str) -> &str {
    if let Some((prefix, _)) = pattern.split_once(['?', '*']) {
        return prefix
            .rsplit_once('/')
            .map(|(prefix, _)| prefix)
            .unwrap_or("");
    };

    pattern
}

/// Match a path against an ant-style pattern
/// Uses glob-match which supports **, *, ?, [ab], {a,b} patterns
pub fn matches(pattern: &str, path: &str) -> bool {
    glob_match::glob_match(pattern, path)
}

/// Check if a path could potentially match a pattern
/// This is used to skip entire directory trees during traversal
/// Returns true if the path matches OR if it's a prefix that could lead to matches
pub fn match_start(pattern: &str, path: &str) -> bool {
    // If it already matches, return true
    if matches(pattern, path) {
        return true;
    }

    // Check if this path is a potential prefix for matches
    let prefix = longest_path_prefix(pattern);

    // If the pattern has no prefix (e.g., "*.xml"), we need to explore everything
    if prefix.is_empty() {
        return true;
    }

    // Check if path is on the way to where matches could occur
    // Either path starts with the pattern's prefix, or the pattern's prefix starts with path
    path.starts_with(prefix) || prefix.starts_with(path)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_longest_path_prefix() {
        assert_eq!(longest_path_prefix("/test/**/*.xml"), "/test");
        assert_eq!(longest_path_prefix("/test/foo.xml"), "/test/foo.xml");
        assert_eq!(longest_path_prefix("**/*.xml"), "");
        assert_eq!(longest_path_prefix("/test/*/foo.xml"), "/test");
    }

    #[test]
    fn test_match_start() {
        // Pattern: /test/**/*.xml
        assert!(match_start("/test/**/*.xml", "/test"));
        assert!(match_start("/test/**/*.xml", "/test/foo"));
        assert!(match_start("/test/**/*.xml", "/test/foo/bar.xml"));
        assert!(!match_start("/test/**/*.xml", "/other"));

        // Pattern: testdata/*/application.properties
        assert!(match_start("testdata/*/application.properties", "testdata"));
        assert!(match_start(
            "testdata/*/application.properties",
            "testdata/subfolder"
        ));
        assert!(match_start(
            "testdata/*/application.properties",
            "testdata/subfolder/application.properties"
        ));
        assert!(!match_start("testdata/*/application.properties", "other"));

        // Pattern with no prefix (**.xml) - should match everything
        assert!(match_start("**/*.xml", "any"));
        assert!(match_start("**/*.xml", "any/path"));
        assert!(match_start("*.xml", "foo.xml"));
    }
}
